Add support for multiple IHC controllers

mopolus opened the pull request:
## Description:
Most IHC systems only have one controller but the system can be setup with a linked secondary controller.
I have updated the code to have it support additional controllers. Existing configuration is not impacted if auto setup is used. Manual configuration must though be moved to the IHC configuration section as in the example below.

**Breaking change**
Support for multiple IHC controllers added. Manual IHC device configuration must be moved from platform sections to controller configuration section under the ihc component section.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7302

## Example entry for `configuration.yaml` (if applicable):
```yaml
ihc:
  - url: 'http://192.168.1.3'
    username: YOUR_USERNAME
    password: YOUR_PASSWORD
    info: true 
  - url: 'http://192.168.1.4'
    username: YOUR_USERNAME2
    password: YOUR_PASSWORD2
    info: true 
    binary_sensor:
      - id: 12345
        name: switch_front_door
        inverting: false
        note: Magnet switch in door
        type: door
      - id: 23451 
      ...
```

## Checklist:
  - [X ] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [X ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) 

If the code communicates with devices, web services, or third-party tools:
  - [] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

mopolus committed:
Added support for secondary IHC controller

Most IHC systems only have one controller but the system can be setup with a linked secondary controller.
I have updated the code to have it support both primary and secondary controller. Existing configuration is not impacted and secondary controller can be setup the same way, with similar settings nested under 'secondary' in the configuration

homeassistant commented:
Hi @mopolus, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#18058).

Once you do that we will be able to review and accept this pull request.

Thanks!

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -26,21 +26,37 @@
+            vol.Required(CONF_USERNAME): cv.string,
+            vol.Required(CONF_PASSWORD): cv.string,
+            vol.Optional(CONF_AUTOSETUP, default=True): cv.boolean,
+            vol.Optional(CONF_INFO, default=True): cv.boolean,    
```
whitespaces removed

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -26,21 +26,37 @@
+            vol.Optional(CONF_AUTOSETUP, default=True): cv.boolean,
+            vol.Optional(CONF_INFO, default=True): cv.boolean,    
+        }),
+}, extra=vol.ALLOW_EXTRA)    
```
whitespaces removed

mopolus committed:
Update __init__.py

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -99,30 +115,41 @@
-    from ihcsdk.ihccontroller import IHCController
     conf = config[DOMAIN]
+    ihc_status = ihc_setup(hass, config, conf, CONTROLLER_ID[PRIMARY])
+    
```
whitespaces removed

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -99,30 +115,41 @@
+    """Setup optional secondary controller."""
+    if ihc_status and CONF_CONTROLLER2 in conf:
+        conf2 = conf[CONF_CONTROLLER2]
+        ihc_status = ihc_setup(hass, config, conf2, CONTROLLER_ID[SECONDARY]) 
```
whitespaces removed

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -99,30 +115,41 @@
+        conf2 = conf[CONF_CONTROLLER2]
+        ihc_status = ihc_setup(hass, config, conf2, CONTROLLER_ID[SECONDARY]) 
+    return ihc_status
+    
```
whitespaces removed

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -99,30 +115,41 @@
+
+def ihc_setup(hass, config, conf, controller_id):
+    from ihcsdk.ihccontroller import IHCController
+    
```
whitespaces removed

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -99,30 +115,41 @@
     url = conf[CONF_URL]
     username = conf[CONF_USERNAME]
     password = conf[CONF_PASSWORD]
+    
```
whitespaces removed

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -99,30 +115,41 @@
 
     if (conf[CONF_AUTOSETUP] and
-            not autosetup_ihc_products(hass, config, ihc_controller)):
+            not autosetup_ihc_products(hass, config, ihc_controller,controller_id)):
```
whitespace & linebreak added

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -99,30 +115,41 @@
+    return True
 
-def autosetup_ihc_products(hass: HomeAssistantType, config, ihc_controller):
+def autosetup_ihc_products(hass: HomeAssistantType, config, ihc_controller, controller_id):
```
line inserted and linebreak added

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -143,14 +170,14 @@ def autosetup_ihc_products(hass: HomeAssistantType, config, ihc_controller):
     for component in IHC_PLATFORMS:
         component_setup = auto_setup_conf[component]
-        discovery_info = get_discovery_info(component_setup, groups)
+        discovery_info = get_discovery_info(component_setup, groups, controller_id)
```
linebreak added

mopolus committed:
Update __init__.py

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -26,21 +26,37 @@
 IHC_INFO = 'info'
 AUTO_SETUP_YAML = 'ihc_auto_setup.yaml'
+CONF_CONTROLLER2 = 'secondary'
+CONTROLLER_ID = [2,1]
```
whitespace added

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -26,21 +26,37 @@
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_AUTOSETUP, default=True): cv.boolean,
+    vol.Optional(CONF_INFO, default=True): cv.boolean,    
```
whitespace removed

mopolus committed:
Update __init__.py

mopolus commented on homeassistant/components/binary_sensor/ihc.py:
```diff
@@ -7,9 +7,10 @@
-from homeassistant.components.ihc import (
-    validate_name, IHC_DATA, IHC_CONTROLLER, IHC_INFO)
-from homeassistant.components.ihc.const import CONF_INVERTING
+from homeassistant.components.ihc import (    
```
whitespaces removed

mopolus commented on homeassistant/components/binary_sensor/ihc.py:
```diff
@@ -32,14 +34,18 @@
+            ihc_key = IHC_DATA.format(ctrl_id)
+            info = hass.data[ihc_key][IHC_INFO]
+            ihc_controller = hass.data[ihc_key][IHC_CONTROLLER]
+            
```
whitespaces removed

mopolus commented on homeassistant/components/binary_sensor/ihc.py:
```diff
@@ -49,6 +55,12 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         for sensor_cfg in binary_sensors:
             ihc_id = sensor_cfg[CONF_ID]
+            # Get controller id
+            ihc_secondary = bool(sensor_cfg[CONF_SECONDARY])            
```
whitespaces removed

mopolus committed:
Update ihc.py

mopolus commented on homeassistant/components/switch/ihc.py:
```diff
@@ -5,9 +5,11 @@
 
-from homeassistant.components.ihc import (
-    validate_name, IHC_DATA, IHC_CONTROLLER, IHC_INFO)
+from homeassistant.components.ihc import (    
```
whitespaces removed

mopolus committed:
Update ihc.py

mopolus commented on homeassistant/components/switch/ihc.py:
```diff
@@ -19,27 +21,37 @@
+            ihc_key = IHC_DATA.format(ctrl_id)
+            info = hass.data[ihc_key][IHC_INFO]
+            ihc_controller = hass.data[ihc_key][IHC_CONTROLLER]
+            
```
whitespaces removed

mopolus commented on homeassistant/components/switch/ihc.py:
```diff
@@ -19,27 +21,37 @@
         for switch in switches:
             ihc_id = switch[CONF_ID]
+            # Get controller id
+            ihc_secondary = bool(switch[CONF_SECONDARY])            
```
whitespaces removed

mopolus committed:
Update ihc.py

mopolus commented on homeassistant/components/light/ihc.py:
```diff
@@ -30,21 +37,31 @@
         for light in lights:
             ihc_id = light[CONF_ID]
+            # Get controller id
+            ihc_secondary = bool(light[CONF_SECONDARY])     
```
whitespaces removed

mopolus reviewed (commented):
Code updated after comments

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -99,30 +115,44 @@
 
 
-def autosetup_ihc_products(hass: HomeAssistantType, config, ihc_controller):
+def autosetup_ihc_products(hass: HomeAssistantType, config, ihc_controller, 
```
whitespace removed

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -143,14 +173,15 @@ def autosetup_ihc_products(hass: HomeAssistantType, config, ihc_controller):
     for component in IHC_PLATFORMS:
         component_setup = auto_setup_conf[component]
-        discovery_info = get_discovery_info(component_setup, groups)
+        discovery_info = get_discovery_info(component_setup, groups, 
```
whitespace removed

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -99,30 +115,44 @@
 
     if (conf[CONF_AUTOSETUP] and
-            not autosetup_ihc_products(hass, config, ihc_controller)):
+            not autosetup_ihc_products(hass, config, ihc_controller, 
```
whitespace removed

mopolus committed:
Update __init__.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus commented on homeassistant/components/sensor/ihc.py:
```diff
@@ -32,14 +34,18 @@
+            ihc_key = IHC_DATA.format(ctrl_id)
+            info = hass.data[ihc_key][IHC_INFO]
+            ihc_controller = hass.data[ihc_key][IHC_CONTROLLER]
+            
```
whitespaces removed

mopolus committed:
Update __init__.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update __init__.py

mopolus committed:
Update ihc.py

mopolus commented on homeassistant/components/light/ihc.py:
```diff
@@ -55,7 +72,7 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 class IhcLight(IHCDevice, Light):
     """Representation of a IHC light.
-
+    
```
whitespaces removed

mopolus commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -98,31 +114,46 @@
+    conf = config[DOMAIN]
+    ihc_status = ihc_setup(hass, config, conf, CONTROLLER_ID[PRIMARY])
+
+    #Setup optional secondary controller.
```
Added missing whitespace

MartinHjelmare commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -26,21 +26,37 @@
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_AUTOSETUP, default=True): cv.boolean,
+    vol.Optional(CONF_INFO, default=True): cv.boolean,
+    vol.Optional(CONF_CONTROLLER2):
```
Is two the max number of controllers? Otherwise this config format doesn't scale so well.

Perhaps a list of controller configs would work?

mopolus replied:
Two controllers is a max (and I agree... otherwise the scaling would be an issue ;-)

MartinHjelmare replied:
Can we make this a list of controller sections and have the first item in the list represent the first controller and the second item the second controller?

mopolus replied:
I thought of this but since the list is limited to two controllers I think an open ended list would be a bit misleading. The two controller setup is already relatively rare and by extending the setup the way I made it, all the existing configurations with one controler (which I believe covers something like 99.x% of the cases) can be left untouched.

mopolus replied:
I couldn't find any examples with a similar "limited number of linked controller" configuration but if there is a praxis and syntax for doing it i can updated the code accordingly.

MartinHjelmare replied:
We can use `cv.ensure_list` to avoid a breaking change.

mopolus replied:
Do you think of something like this:
```python
IHC_SCHEMA = vol.Schema({
    vol.Required(CONF_URL): cv.string,
    vol.Required(CONF_USERNAME): cv.string,
    vol.Required(CONF_PASSWORD): cv.string,
    vol.Optional(CONF_NAME, default=""): cv.string,
    vol.Optional(CONF_AUTOSETUP, default=True): cv.boolean,
    vol.Optional(CONF_INFO, default=True): cv.boolean,    
}, extra=vol.ALLOW_EXTRA)    

CONFIG_SCHEMA = vol.Schema({
    DOMAIN: vol.Schema({
        vol.Optional(CONF_CONTROLLERS):
            vol.All(cv.ensure_list, [IHC_SCHEMA]),
    }),
}, extra=vol.ALLOW_EXTRA)
```
That would allow for multiple parallel controllers to be setup
```yaml
ihc:
  controllers:
    - url: 'http://192.168.1.3'
      username: YOUR_USERNAME
      password: YOUR_PASSWORD
      info: true 
      secondary: 
    - url: 'http://192.168.1.4'
      username: YOUR_USERNAME2
      password: YOUR_PASSWORD2
      info: true 
```
It would though be a breaking change?

MartinHjelmare replied:
We can avoid a breaking change I think.
```py
CONFIG_SCHEMA = vol.Schema({
    DOMAIN: vol.Schema(vol.All(
        cv.ensure_list,
        [IHC_SCHEMA]))})
```

mopolus committed:
Update __init__.py

mopolus committed:
Update const.py

mopolus committed:
Update ihc.py

MartinHjelmare commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -114,15 +105,20 @@
+    conf = config.get(DOMAIN)
+    index = 0
+    # Set up the IHC component
+    for controller_conf in conf:
```
Use enumerate to get the index while iterating.
```py
for idx, controller_conf in enumerate(conf):
```

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

MartinHjelmare commented on homeassistant/components/binary_sensor/ihc.py:
```diff
@@ -49,9 +55,13 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         for sensor_cfg in binary_sensors:
             ihc_id = sensor_cfg[CONF_ID]
+            controller_id = sensor_cfg[CONF_CONTROLLER]
+            ihc_controller = hass.data[controller_id][IHC_CONTROLLER]
```
This is very brittle. It's easy for the user to make a mistake when naming the controllers which would break this.

Maybe we should make a breaking change by moving the platform configs to the component under the controller section? That would make it easier for the users and we could better guard against breakage.

MartinHjelmare replied:
Yes! What do you think about that?

mopolus replied:
Well.. mostly good. It will be more "clean" and remove the need of naming the controllers and keeping the configuration together. 
My biggest (only) worry is if we break any conventions on configuring binary_sensor, light, switch etc?

mopolus replied:
(and it being a breaking change ;-)

MartinHjelmare replied:
We now recommend to keep all configuration in the component section, so the change would be a step in the right direction in my opinion.

This will be a bigger change than initially planned. Is that ok with you time wise?

mopolus replied:
Great to hear that it is the direction things are moving ;-)
I'm okay time wise with changing it. Thanks for asking!

mopolus replied:
Is there any projects already following the new conventions?
It would be great with a reference in order to slice the schemas correctly

MartinHjelmare replied:
Yes, look at octoprint:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/octoprint.py
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/binary_sensor/octoprint.py

mopolus replied:
Give it a look. I have moved the device configuration to the component section and extended options for manual configuration. I have not yet updated the documentation but will do as soon as you have confirmed. The controllers and devices can now be configured like below:
```yaml
ihc:
  - url: 'http://192.168.1.3'
    username: YOUR_USERNAME
    password: YOUR_PASSWORD
    info: true 
    binary_sensor:
      38490: 
        name: switch_front_door
        inverting: false
        note: Magnet switch in door
        type: door
     64819: 
     ...
  - url: 'http://192.168.1.4'
     ...
```

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update __init__.py

mopolus committed:
Update __init__.py

mopolus committed:
Update __init__.py

mopolus committed:
Update const.py

mopolus committed:
Update __init__.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update __init__.py

mopolus committed:
Update __init__.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update __init__.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

MartinHjelmare commented on homeassistant/components/binary_sensor/ihc.py:
```diff
@@ -3,60 +3,39 @@
-    info = hass.data[IHC_DATA][IHC_INFO]
     devices = []
-    if discovery_info:
+    if discovery_info:  
```
Move this to the beginning of the function body, invert the check and return if true:
```py
if discover_info is None:
    return
```
We will save indentation that way.

mopolus committed:
Update __init__.py

MartinHjelmare commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -98,35 +166,76 @@
+                    }
+                }
+                discovery_info = {sensor_cfg[CONF_NAME]: device}
+                discovery.load_platform(hass, 'binary_sensor', DOMAIN,
```
Don't hardcode to `'binary_sensor'` component. Use `component` from iterator.

mopolus replied:
Doh... How did I missed that.

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

MartinHjelmare commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -9,38 +9,106 @@
+})
+
+BINARY_SENSOR_SCHEMA = vol.Schema({
+    vol.All(cv.ensure_list, [
```
The schema doesn't look right. Looks like it's too many brackets. We want to validate a list of dictionaries.

mopolus replied:
I might have placed it wrong. The intend was to move all the standard settings to DEVICE_SCHEMA and include it in the device schemas. Ie as an alternative to this:
```yaml
BINARY_SENSOR_SCHEMA = vol.Schema({
    vol.All(cv.ensure_list, [
        vol.All({
            vol.Required(CONF_ID): cv.positive_int,
            vol.Optional(CONF_NAME): cv.string,
            vol.Optional(CONF_POSITION): cv.string,
            vol.Optional(CONF_NOTE): cv.string,
            vol.Optional(CONF_TYPE): DEVICE_CLASSES_SCHEMA,
            vol.Optional(CONF_INVERTING, default=False): cv.boolean,
        }, validate_name)
    ])
})
```
What would be the correct way to do the split/include of DEVICE_SCHEMA ?

MartinHjelmare replied:
We can use `.extend(...)` on dictionary based schemas to update them with more items.

mopolus replied:
Thanks. I have updated. Will run some configuration tests tomorrow

mopolus replied:
Help needed! I have splittet the schemas and and is extending them for each sensor. So far so good. 
```python
DEVICE_SCHEMA = vol.Schema({
    vol.Required(CONF_ID): cv.positive_int,
    vol.Optional(CONF_NAME): cv.string,
    vol.Optional(CONF_POSITION): cv.string,
    vol.Optional(CONF_NOTE): cv.string
}, extra=vol.ALLOW_EXTRA)

BINARY_SENSOR_SCHEMA = DEVICE_SCHEMA.extend({
    vol.Optional(CONF_TYPE): DEVICE_CLASSES_SCHEMA,
    vol.Optional(CONF_INVERTING, default=False): cv.boolean,
})
```
My trouble starts when adding them to the main configuration schema.
```python
IHC_SCHEMA = vol.Schema({
    vol.Required(CONF_URL): cv.string,
   ...
    vol.Optional(CONF_BINARY_SENSORS, default=[]):
        vol.All(cv.ensure_list, [BINARY_SENSOR_SCHEMA]),
or
        vol.All(cv.ensure_list, [
            vol.All({
                vol.All([BINARY_SENSOR_SCHEMA])
            }, validate_name)
        ]),
```
The code works but i've have lost the validation in each cases :-(
The original code was
```
PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
    vol.Optional(CONF_BINARY_SENSORS, default=[]):
        vol.All(cv.ensure_list, [
            vol.All({
                vol.Required(CONF_ID): cv.positive_int,
                vol.Optional(CONF_NAME): cv.string,
               ...
            }, validate_name)
        ])
})
```
Suggestions?

mopolus committed:
Update __init__.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update ihc.py

mopolus committed:
Update __init__.py

mopolus committed:
Update __init__.py

mopolus committed:
Update ihc.py

mopolus committed:
Update __init__.py

mopolus committed:
Update __init__.py

mopolus committed:
Update __init__.py

mopolus committed:
Update __init__.py

MartinHjelmare commented:
Please add a paragraph to the PR description, describing the breaking change, that we can copy to the release notes.

MartinHjelmare commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -9,38 +9,108 @@
+
+
+SWITCH_SCHEMA = DEVICE_SCHEMA.extend({
+    vol.Required(CONF_ID): cv.positive_int,
```
These already exist in the `DEVICE_SCHEMA`.

MartinHjelmare commented on homeassistant/components/ihc/__init__.py:
```diff
@@ -98,35 +168,76 @@
+                    }
+                }
+                discovery_info = {sensor_cfg[CONF_NAME]: device}
+                discovery.load_platform(hass, component, DOMAIN,
```
Only call `load_platform` once per component. Collect all the discovery_info in a container and pass that as discovery_info. Iterate the container in the platform to create all entities. Then add all entities in one call to `add_entities`.

mopolus committed:
Update __init__.py

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Can be merged when build passes and a breaking change paragraph is added to the PR description.

mopolus commented:
> Can be merged when build passes and a breaking change paragraph is added to the PR description.

Breaking change paragraph added

MartinHjelmare commented:
Good! I tweaked it a bit.

mopolus committed:
Update __init__.py

indentation was incorrect for "load_platform" in "get_manual_configuration". Load_platform was not called with the correct component name

MartinHjelmare reviewed (approved):
Good!

YoinkZed commented:
Dear Mopulos,
It looks like we are a few that are experience some problems after the latest changes.
When we add our lights, sensors and switches manually, they don't show up (at least some of them don't for some people and for others they do).
If I do the auto_setup it works like a charm, but could you perhaps have a look or do you have an idea why?

https://community.home-assistant.io/t/ihc-problems-light-not-present-after-update/84999/9

MartinHjelmare commented:
There's already an issue for this problem.

Merged PRs should not be used for support or bug reports. Thanks!
